src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,276 @@
+                        var args = new PinValueChangedEventArgs(eventType, eventHandler.PinNumber);
+                        eventHandler?.OnPinValueChanged(args, eventType);
+                    }
+                }
+            });
+        }
+
+        protected internal override void ClosePin(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                Interop.ReleaseGpiodLine(pin);
+                _pinNumberToSafeLineHandle.Remove(pinNumber);
+            }
+        }
+
+        protected internal override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => throw new
+            PlatformNotSupportedException("This driver is generic so it cannot perform conversions between pin numbering schemes.");
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                return (Interop.GetLineDirection(pin) == 1) ? PinMode.Input : PinMode.Output;
+            }
+            return PinMode.Input;
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod Api seems do not support pull up or pull down resistors.
+            return mode != PinMode.InputPullDown && mode != PinMode.InputPullUp;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin == null)
+            {
+                throw new IOException($"Pin number {pinNumber} not available for chip: {_chip} error: {Marshal.GetLastWin32Error()}");
+            }
+            _pinNumberToSafeLineHandle.Add(pinNumber, pin);
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                int value = Interop.GetGpiodLineValue(pin);
+                if (value == -1)
+                {
+                    throw new IOException($"Error while reading value from pin number: {pinNumber}, error: {Marshal.GetLastWin32Error()}");
+                }
+                return (value == 1) ? PinValue.High : PinValue.Low;
+            }
+            return PinValue.Low;
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            if (_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                eventHandler.ValueFalling -= callback;
+                eventHandler.ValueRising -= callback;
+                if (eventHandler.IsCallbackListEmpty())
+                {
+                    eventHandler.CancellationTokenSource.Cancel();
```

Comment:
I believe in here you want to make sure that after sending the cancellation, you want to validate that the task is in fact not running any longer before you dispose it or release the handle.

Reply:
Good point, is there any way to validate if the task is running without a reference to the task?  Might be stupid question, most likely not 🤔

Reply:
task.IsCompleted, you will need reference though (also you should have it for the Dispose)

Reply:
in here you probably want to Wait though

Reply:
yeah I saw that task.IsCompleted method, my question was without referencing the task object. As for now am not keeping reference to the task